kernel/src/scheduler.rs

```diff
@@ -16,52 +16,74 @@ use crate::process::StoppedExecutingReason;
 
 /// Trait which any scheduler must implement.
 pub trait Scheduler<C: Chip> {
-    /// Decide which process to run next.
+    /// Determine if the kernel should handle interrupts
     ///
-    /// The scheduler must decide whether to run a process, and if so, which
-    /// one. If the scheduler chooses not to run a process, it can request that
-    /// the chip enter sleep mode.
+    /// The default implementation always returns true if there are any pending interrupts.
+    /// Custom schedulers may wish to reimplement this method to defer interrupt handling.
+    fn should_kernel_handle_interrupts(&self, chip: &C) -> bool {
+        chip.has_pending_interrupts()
+    }
+
+    /// Determine if the kernel should handle a deferred call
     ///
-    /// If the scheduler selects a process to run it must provide its `ProcessId`
-    /// and an optional timeslice length in microseconds to provide to that
-    /// process. If the timeslice is `None`, the process will be run
-    /// cooperatively (i.e. without preemption). Otherwise the process will run
-    /// with a timeslice set to the specified length.
-    fn next(&self) -> SchedulingDecision;
+    /// The default implementation always returns true if there are any pending deferred calls.
+    /// Custom schedulers may wish to reimplement this method to defer deferred calls.
+    fn should_kernel_handle_deferred_call(&self) -> bool {
+        DeferredCall::has_tasks()
+    }
 
-    /// Inform the scheduler of why the last process stopped executing, and how
-    /// long it executed for. Notably, `execution_time_us` will be `None`
-    /// if the the scheduler requested this process be run cooperatively.
-    fn result(&self, result: StoppedExecutingReason, execution_time_us: Option<u32>);
+    /// Ask the scheduler whether to take a break from executing userspace
+    /// processes to handle kernel tasks.
+    ///
+    /// The default implementation prioritises interrupts over deferred calls. Custom scheduler
+    /// may reimplement this method to prioritise deferred calls over interrupts.
+    fn should_kernel_do_work(&self, chip: &C) -> Option<KernelWorkType> {
+        if self.should_kernel_handle_interrupts(chip) {
+            return Some(KernelWorkType::Interrupts);
+        }
+
+        if self.should_kernel_handle_deferred_call() {
+            return Some(KernelWorkType::DeferredCalls);
+        }
+
+        None
+    }
+
+    /// Determine the next process to run. If no suitable process is found, return None. If a
+    /// suitable process is found, return its identifier and its assigned time slice in
+    /// microseconds. If no time slice is assigned, the process is meant to run cooperatively.
+    fn next_process(&self) -> Option<(ProcessId, Option<u32>)>;
```

Comment:
`next_process()` explicitly states that the method should decide which process to run next. Returning an `Option` indicates that no process is able to run.